### PR TITLE
feat: Update Insight category, document sentiment and archive insights

### PR DESCRIPTION
**Pull request by jobenjada**
solves https://github.com/formbricks/internal/issues/402

change both sentiment and category if AI got it wrong :) 

Open questions:

Problem: If I update the category of an insight in a survey, I need to invalidate the surveyCache somehow for the change to reflect on the summary page. Right now, we need to wait for the interval to invalidate the cache.  How to do that manually?

Problem: The documents in the insight-sheet are flashing a lot. Would be cooler if they werent, tried debouncing but didn't work well. Also, quite often there are duplicates.

General question: Do we need to secure the function which runs on the server? Or check for permissions? Or is it enough to do that via the actions?

## Summary by CodeRabbit

- **New Features**
	- Introduced a `BadgeSelect` component for customizable dropdown menus.
	- Added a `SentimentSelect` component for selecting sentiment options.
	- Implemented an `updateDocumentAction` for document updates with authorization checks.
	- Enhanced the `updateInsightAction` for updating insights with improved error handling.

- **Styling Changes**
	- Updated various components' styles for consistency, changing background and text colors from gray to slate.
	- Modified the `Loading` component's placeholders for a more uniform appearance.
	- Adjusted button sizes and styles in multiple components for better visual consistency.
	- Changed text colors in several components to improve visual presentation.

- **Text Updates**
	- Revised localization strings for clarity and consistency across multiple languages, including updates to specific user-facing messages.

- **Bug Fixes**
	- Improved error handling in several components to enhance user experience.

- **Documentation**
	- Updated documentation to reflect new components and functionalities.

### CODE REVIEW

File: apps/web/modules/ee/insights/components/insight-sheet/lib/documents.ts [hunk 1]
```diff
@@ -124,3 +124,57 @@
...
+): Promise<TDocument> => {
+  try {
+    const updatedDocument = await prisma.document.update({
+      where: { id: documentId },
+      data: updates,
+    });
+    documentCache.revalidate({ environmentId: environmentId });
+    if (insightId) {
+      insightCache.revalidate({ id: insightId });
+    }
```
[the remark this answers is not in the document]
Comment (jobenjada) [reply apps/web/modules/ee/insights/components/insight-sheet/lib/documents.ts:171]: agreed but this was kinda neccessary due to how the caching is setup, it didnt work with document cache invalidation

File: apps/web/modules/ee/insights/components/insight-sheet/lib/documents.ts [hunk 2]
```diff
@@ -124,3 +124,57 @@
...
+
+          throw error;
+        }
+      },
+      [`getDocumentById-${documentId}`],
+      {
+        tags: [documentCache.tag.byId(documentId)],
+      }
+    )()
+);
```
[the remark this answers is not in the document]
Comment (jobenjada) [reply apps/web/modules/ee/insights/components/insight-sheet/lib/documents.ts:155]: should I do this @mattinannt  ?

File: apps/web/modules/ee/insights/components/insight-sheet/index.tsx [hunk 3]
```diff
@@ -42,7 +45,7 @@ export const InsightSheet = ({
 }: InsightSheetProps) => {
   const [documents, setDocuments] = useState<TDocument[]>([]);
   const [page, setPage] = useState(1);
-  const [hasMore, setHasMore] = useState(true);
+  const [hasMore, setHasMore] = useState(false); // Default to false
```
[the remark this answers is not in the document]
Comment (jobenjada) [reply apps/web/modules/ee/insights/components/insight-sheet/index.tsx:48]: done

File: apps/web/modules/ee/insights/components/insight-sheet/index.tsx [hunk 4]
```diff
@@ -112,44 +116,37 @@
...
-                    <Badge text="Neutral" size="tiny" type="gray" />
-                  ) : document.sentiment === "negative" ? (
-                    <Badge text="Negative" size="tiny" type="error" />
-                  ) : null}
+                  <SentimentSelect
+                    documentId={document.id}
+                    insightId={insight.id}
+                    sentiment={document.sentiment}
+                    environmentId={environmentId}
+                  />
```
[the remark this answers is not in the document]
Comment (jobenjada) [reply apps/web/modules/ee/insights/components/insight-sheet/index.tsx:149]: works well

File: packages/ui/components/Badge/index.tsx
Comment (mattinannt) [inline packages/ui/components/Badge/index.tsx:1]: The badge is a simple UI component ( https://ui.shadcn.com/docs/components/badge ) that also works as a server component. With your change you have made it much more complex and also forced it to be a client component due to the dropdown.
Please add a new component instead.

File: apps/web/modules/ee/insights/experience/types/stats.ts [hunk 5]
```diff
@@ -2,7 +2,7 @@ import { z } from "zod";
 
 export const ZStats = z.object({
   sentimentScore: z.number().optional(),
-  overallSentiment: z.enum(["positive", "negative"]).optional(),
+  overallSentiment: z.enum(["positive", "negative", "balanced"]).optional(),
```
Comment (mattinannt) [inline apps/web/modules/ee/insights/experience/types/stats.ts:5]: I think `neutral` is much more common for sentiment description.

File: apps/web/modules/ee/insights/experience/components/stats.tsx [hunk 6]
```diff
@@ -79,20 +87,21 @@ export const ExperiencePageStats = ({ statsFrom, environmentId }: ExperiencePage
...
-              )}
-            </div>
+          <CardContent className={cn(isLoading && "animate-pulse")}>
             {stat.key === "sentimentScore" && stats.overallSentiment && (
-              <div>
+              <div className="flex items-center font-medium text-slate-700">
                 {stats.overallSentiment === "positive" ? (
-                  <Badge text="Positive" type="success" size="tiny" />
+                  <TooltipRenderer tooltipContent="Mostly positive">
+                    <SmileIcon className="h-10 w-10" strokeWidth={1.5} />
```
Comment (mattinannt) [inline apps/web/modules/ee/insights/experience/components/stats.tsx:95]: Screenshot 2024-10-29 at 06 13 58

We lose a lot of information by removing the number. When comparing sentiment between two time periods we don't know if it's going up or down if we just see the same smile/category. Please discuss with chatGPT what the best scientific approach is to display the development of sentiment over time in a dashboard.

File: apps/web/modules/ee/insights/experience/components/stats.tsx
Comment (mattinannt) [inline apps/web/modules/ee/insights/experience/components/stats.tsx:1]: Screenshot 2024-10-29 at 06 17 33

All stats except the sentiment are broken in my test.

File: packages/database/schema.prisma [hunk 7]
```diff
@@ -673,6 +673,7 @@ model Insight {
   description      String
   vector           Unsupported("vector(512)")?
   documentInsights DocumentInsight[]
+  archived         Boolean                     @default(false)
```
Comment (mattinannt) [inline packages/database/schema.prisma:676]: rename to `isArchived` to follow new naming convention

Comment (mattinannt) [inline packages/database/schema.prisma:676]: I think the 'archive' approach needs more refinement. Currently insights don't reappear. It's not ideal if someone complained about app performance a year ago, I archived it, and now people are complaining again (even in a different survey) and I don't see it in the environment insights.

File: packages/types/badge.ts
Comment (mattinannt) [inline packages/types/badge.ts:1]: Please store ui component types directly in the ui component.

File: packages/types/insights.ts [hunk 8]
```diff
@@ -16,6 +16,7 @@ export const ZInsight = z.object({
   _count: z.object({
     documentInsights: z.number(),
   }),
+  archived: z.boolean().optional(),
```
Comment (mattinannt) [inline packages/types/insights.ts:19]: Since it's always present in the database it shouldn't be `optional`, right?

**Review by mattinannt — CHANGES_REQUESTED** [review]
@jobenjada thanks for the PR 🚀 I added a few comments and change requests 😊

I think for the future it's better to split PRs like this into several smaller PRs so that we can merge design changes, for example, fairly quickly. Introducing completely new features can lead to a lot of iterations within the PR or even the PR being rejected.

**Review by jobenjada — COMMENTED** [review]
as discussed

**Review by mattinannt — APPROVED** [review]

**Pull request merged**